Zend/zend_compile.c

```diff
@@ -2974,7 +2988,234 @@ void zend_compile_static_call(znode *result, zend_ast *ast, uint32_t type) /* {{
+	zend_ast *extends_ast = decl->child[0];
+	zend_ast *implements_ast = decl->child[1];
+	zend_ast *stmt_ast = decl->child[2];
+
+	zend_string *name = decl->name, *lcname, *import_name = NULL;
+	zend_class_entry *ce = zend_arena_alloc(&CG(arena), sizeof(zend_class_entry));
+	zend_op *opline;
+	znode declare_node, extends_node;
+	zend_class_entry *active = CG(active_class_entry);
+	
+	if (decl->flags & ZEND_ACC_ANON_CLASS) {
+	    name = 
+	        zend_name_anon_class((zend_ast*)name TSRMLS_CC);
+
+	    /* do not support serial classes */
+	    ce->serialize = zend_class_serialize_deny;
+	    ce->unserialize = zend_class_unserialize_deny;
+	}
+	
+	if (CG(active_class_entry) && !((decl->flags & ZEND_ACC_ANON_CLASS) == ZEND_ACC_ANON_CLASS)) {
+            zend_error(E_COMPILE_ERROR, "Class declarations may not be nested");
+            return NULL;
+        }
+
+	if (ZEND_FETCH_CLASS_DEFAULT != zend_get_class_fetch_type(name)) {
+		zend_error_noreturn(E_COMPILE_ERROR, "Cannot use '%s' as class name as it is reserved",
+			name->val);
+	}
+    
+    lcname = zend_string_alloc(name->len, 0);
+	zend_str_tolower_copy(lcname->val, name->val, name->len);
+	
+
+	if (CG(current_import)) {
+		import_name = zend_hash_find_ptr(CG(current_import), lcname);
+	}
+
+	if (CG(current_namespace)) {
+		name = zend_prefix_with_ns(name TSRMLS_CC);
+
+		zend_string_release(lcname);
+		lcname = zend_string_alloc(name->len, 0);
+		zend_str_tolower_copy(lcname->val, name->val, name->len);
+	} else {
+		zend_string_addref(name);
+	}
+
+	if (import_name && !zend_string_equals_str_ci(lcname, import_name)) {
+		zend_error_noreturn(E_COMPILE_ERROR, "Cannot declare class %s "
+			"because the name is already in use", name->val);
+	}
+
+	name = zend_new_interned_string(name TSRMLS_CC);
+	lcname = zend_new_interned_string(lcname TSRMLS_CC);
+
+	ce->type = ZEND_USER_CLASS;
+	ce->name = name;
+	zend_initialize_class_data(ce, 1 TSRMLS_CC);
+
+	ce->ce_flags |= decl->flags;
+	ce->info.user.filename = zend_get_compiled_filename(TSRMLS_C);
+	ce->info.user.line_start = decl->start_lineno;
+	ce->info.user.line_end = decl->end_lineno;
+	if (decl->doc_comment) {
+		ce->info.user.doc_comment = zend_string_copy(decl->doc_comment);
+	}
+
+	if (extends_ast) {
+		if (ZEND_CE_IS_TRAIT(ce)) {
+			zend_error_noreturn(E_COMPILE_ERROR, "A trait (%s) cannot extend a class. "
+				"Traits can only be composed from other traits with the 'use' keyword. Error",
+				name->val);
+		}
+
+		if (!zend_is_const_default_class_ref(extends_ast)) {
+			zend_string *extends_name = zend_ast_get_str(extends_ast);
+			zend_error_noreturn(E_COMPILE_ERROR,
+				"Cannot use '%s' as class name as it is reserved", extends_name->val);
+		}
+
+		zend_compile_class_ref(&extends_node, extends_ast TSRMLS_CC);
+	}
+
+	opline = get_next_op(CG(active_op_array) TSRMLS_CC);
+	zend_make_var_result(&declare_node, opline TSRMLS_CC);
+
+	// TODO.AST drop this
+	GET_NODE(&CG(implementing_class), opline->result);
```

Comment:
It's likely that this TODO will be problematic for you - please add a test case where an anon class is defined inside another class and then *after* the anon class use some trait in the outer class. I suspect the trait will be added to the wrong class.